File: src/robot/robot/stereo_splitter_node.py
#!/usr/bin/env python3
"""拆分横向拼接的 UVC 双目图像，并发布左右图像和标定信息。"""

import os

import numpy as np
import rclpy
import yaml
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import CameraInfo, Image


class StereoSplitter(Node):
    """保持同一采集时间戳，将一帧横向拼接图像拆成左右两帧。"""

    def __init__(self):
        super().__init__('stereo_splitter')
        self.declare_parameter('input_topic', '/stereo/image_raw')
        self.declare_parameter('left_image_topic', '/stereo/left/image_raw')
        self.declare_parameter('right_image_topic', '/stereo/right/image_raw')
        self.declare_parameter('left_camera_info_topic', '/stereo/left/camera_info')
        self.declare_parameter('right_camera_info_topic', '/stereo/right/camera_info')
        self.declare_parameter('left_frame_id', 'stereo_left_optical_frame')
        self.declare_parameter('right_frame_id', 'stereo_right_optical_frame')
        self.declare_parameter('left_first', True)
        self.declare_parameter('frame_skip', 0)
        self.declare_parameter('output_encoding', 'passthrough')
        self.declare_parameter('calibration_mode', False)
        self.declare_parameter('left_calibration_file', '')
        self.declare_parameter('right_calibration_file', '')

        self.left_first = bool(self.get_parameter('left_first').value)
        self.frame_skip = max(0, int(self.get_parameter('frame_skip').value))
        self.output_encoding = str(self.get_parameter('output_encoding').value)
        self.calibration_mode = bool(self.get_parameter('calibration_mode').value)
        self.left_frame = str(self.get_parameter('left_frame_id').value)
        self.right_frame = str(self.get_parameter('right_frame_id').value)
        self.frame_count = 0

        self.left_info = self._load_camera_info(
            str(self.get_parameter('left_calibration_file').value),
            self.left_frame,
            'left',
        )
        self.right_info = self._load_camera_info(
            str(self.get_parameter('right_calibration_file').value),
            self.right_frame,
            'right',
        )

        input_topic = str(self.get_parameter('input_topic').value)
        self.left_pub = self.create_publisher(
            Image, str(self.get_parameter('left_image_topic').value),
            qos_profile_sensor_data)
        self.right_pub = self.create_publisher(
            Image, str(self.get_parameter('right_image_topic').value),
            qos_profile_sensor_data)
        self.left_info_pub = self.create_publisher(
            CameraInfo, str(self.get_parameter('left_camera_info_topic').value),
            qos_profile_sensor_data)
        self.right_info_pub = self.create_publisher(
            CameraInfo, str(self.get_parameter('right_camera_info_topic').value),
            qos_profile_sensor_data)
        self.create_subscription(
            Image, input_topic, self.image_callback, qos_profile_sensor_data)
        self.get_logger().info(
            f'双目拆分已启动: {input_topic}, left_first={self.left_first}, '
            f'frame_skip={self.frame_skip}')

    def _load_camera_info(self, path, frame_id, side):
        """读取 camera_calibration 生成的 YAML；标定模式允许文件缺失。"""
        if not path or not os.path.isfile(path):
            message = f'{side} 标定文件不存在: {path or "(空)"}'
            if self.calibration_mode:
                self.get_logger().warn(message + '；标定模式仅发布原始图像')
                return None
            self.get_logger().error(message + '；请先完成双目标定')
            return None

        try:
            with open(path, 'r', encoding='utf-8') as stream:
                data = yaml.safe_load(stream)
            info = CameraInfo()
            info.width = int(data['image_width'])
            info.height = int(data['image_height'])
            info.distortion_model = str(data.get(
                'distortion_model', 'plumb_bob'))
            info.d = [float(value) for value in
                      data['distortion_coefficients']['data']]
            info.k = [float(value) for value in data['camera_matrix']['data']]
            info.r = [float(value) for value in
                      data['rectification_matrix']['data']]
            info.p = [float(value) for value in
                      data['projection_matrix']['data']]
            info.header.frame_id = frame_id
            if not self.calibration_mode:
                if info.k[0] <= 0.0 or info.k[4] <= 0.0:
                    self.get_logger().error(
                        f'{side} 标定文件仍是模板，不能启动深度处理')
                    return None
                if side == 'right' and abs(info.p[3]) <= 1e-9:
                    self.get_logger().error(
                        'right 投影矩阵没有有效 Tx，不能用标称基线替代标定')
                    return None
            return info
        except (KeyError, TypeError, ValueError, yaml.YAMLError) as exc:
            self.get_logger().error(f'{side} 标定文件格式错误: {exc}')
            return None

    @staticmethod
    def _bytes_per_pixel(encoding):
        return {
            'mono8': 1,
            '8UC1': 1,
            'rgb8': 3,
            'bgr8': 3,
            'rgba8': 4,
            'bgra8': 4,
        }.get(encoding)

    def image_callback(self, msg):
        self.frame_count += 1
        if (self.frame_count - 1) % (self.frame_skip + 1):
            return
        if msg.width < 2 or msg.width % 2:
            self.get_logger().error(
                f'拼接图宽度必须是正偶数，当前为 {msg.width}',
                throttle_duration_sec=2.0)
            return

        bytes_per_pixel = self._bytes_per_pixel(msg.encoding)
        if bytes_per_pixel is None:
            self.get_logger().error(
                f'暂不支持输入编码 {msg.encoding}；请让 usb_cam 输出 '
                'mono8/rgb8/bgr8/rgba8/bgra8',
                throttle_duration_sec=2.0)
            return

        half_width = msg.width // 2
        row_bytes = msg.width * bytes_per_pixel
        if msg.step < row_bytes or len(msg.data) < msg.step * msg.height:
            self.get_logger().error('输入 Image 的 step 或 data 长度无效')
            return

        rows = np.ndarray(
            shape=(msg.height, msg.step),
            dtype=np.uint8,
            buffer=msg.data,
        )
        packed = rows[:, :row_bytes].reshape(
            msg.height, msg.width, bytes_per_pixel)
        first = np.ascontiguousarray(packed[:, :half_width])
        second = np.ascontiguousarray(packed[:, half_width:])
        left, right = (first, second) if self.left_first else (second, first)

        encoding = (msg.encoding if self.output_encoding == 'passthrough'
                    else self.output_encoding)
        if encoding != msg.encoding:
            self.get_logger().error(
                '当前节点不做颜色空间转换；output_encoding 应设为 passthrough '
                '或与输入编码一致',
                throttle_duration_sec=2.0)
            return

        left_msg = self._make_image(msg, left, half_width, self.left_frame)
        right_msg = self._make_image(msg, right, half_width, self.right_frame)
        self.left_pub.publish(left_msg)
        self.right_pub.publish(right_msg)
        self._publish_info(self.left_info_pub, self.left_info, left_msg)
        self._publish_info(self.right_info_pub, self.right_info, right_msg)

    @staticmethod
    def _make_image(source, array, width, frame_id):
        output = Image()
        output.header.stamp = source.header.stamp
        output.header.frame_id = frame_id
        output.height = source.height
        output.width = width
        output.encoding = source.encoding
        output.is_bigendian = source.is_bigendian
        output.step = width * array.shape[2]
        output.data = array.tobytes()
        return output

    @staticmethod
    def _publish_info(publisher, template, image):
        if template is None:
            return
        template.header.stamp = image.header.stamp
        template.header.frame_id = image.header.frame_id
        publisher.publish(template)


def main(args=None):
    rclpy.init(args=args)
    node = StereoSplitter()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/robot_stereo_components/src/stereo_splitter_node.cpp
#include <algorithm>
#include <chrono>
#include <cstddef>
#include <cstdint>
#include <cstring>
#include <functional>
#include <memory>
#include <stdexcept>
#include <string>
#include <utility>

#include "camera_info_manager/camera_info_manager.hpp"
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/camera_info.hpp"
#include "sensor_msgs/msg/image.hpp"

namespace robot_stereo_components
{

class StereoSplitterNode final : public rclcpp::Node
{
public:
  StereoSplitterNode()
  : Node("stereo_splitter")
  {
    input_topic_ = declare_parameter<std::string>("input_topic", "/stereo/image_raw");
    left_topic_ = declare_parameter<std::string>(
      "left_image_topic", "/stereo/left/image_raw");
    right_topic_ = declare_parameter<std::string>(
      "right_image_topic", "/stereo/right/image_raw");
    left_info_topic_ = declare_parameter<std::string>(
      "left_camera_info_topic", "/stereo/left/camera_info");
    right_info_topic_ = declare_parameter<std::string>(
      "right_camera_info_topic", "/stereo/right/camera_info");
    left_frame_ = declare_parameter<std::string>(
      "left_frame_id", "stereo_left_optical_frame");
    right_frame_ = declare_parameter<std::string>(
      "right_frame_id", "stereo_right_optical_frame");
    left_first_ = declare_parameter<bool>("left_first", true);
    calibration_mode_ = declare_parameter<bool>("calibration_mode", false);
    recognition_rate_ = declare_parameter<double>("recognition_rate", 10.0);
    const auto left_calibration = declare_parameter<std::string>(
      "left_calibration_file", "");
    const auto right_calibration = declare_parameter<std::string>(
      "right_calibration_file", "");

    left_camera_info_ = std::make_unique<camera_info_manager::CameraInfoManager>(
      this, "stereo_left", to_url(left_calibration));
    right_camera_info_ = std::make_unique<camera_info_manager::CameraInfoManager>(
      this, "stereo_right", to_url(right_calibration));
    if (!calibration_mode_ &&
      (!left_camera_info_->isCalibrated() || !right_camera_info_->isCalibrated()))
    {
      throw std::runtime_error("双目标定文件未正确加载，拒绝发布未标定图像");
    }

    // 大图输入采用传感器 QoS；图像和 CameraInfo 使用完全相同的可靠单帧队列，
    // 避免处理链繁忙时 CameraInfo 独自积压并触发 image_proc 不同步告警。
    const auto input_qos = rclcpp::SensorDataQoS().keep_last(1);
    const auto output_qos = rclcpp::QoS(rclcpp::KeepLast(1)).reliable();
    left_image_pub_ = create_publisher<sensor_msgs::msg::Image>(left_topic_, output_qos);
    right_image_pub_ = create_publisher<sensor_msgs::msg::Image>(right_topic_, output_qos);
    left_info_pub_ = create_publisher<sensor_msgs::msg::CameraInfo>(
      left_info_topic_, output_qos);
    right_info_pub_ = create_publisher<sensor_msgs::msg::CameraInfo>(
      right_info_topic_, output_qos);
    subscription_ = create_subscription<sensor_msgs::msg::Image>(
      input_topic_, input_qos,
      std::bind(&StereoSplitterNode::image_callback, this, std::placeholders::_1));
    RCLCPP_INFO(get_logger(), "C++ 双目拆分器已启动，输入 %s", input_topic_.c_str());
  }

private:
  static std::string to_url(const std::string & path)
  {
    if (path.empty() || path.find("://") != std::string::npos) {
      return path;
    }
    return "file://" + path;
  }

  void image_callback(const sensor_msgs::msg::Image::ConstSharedPtr msg)
  {
    if (msg->width < 2 || msg->width % 2 != 0 || msg->height == 0 || msg->step == 0) {
      RCLCPP_ERROR_THROTTLE(
        get_logger(), *get_clock(), 2000,
        "拼接图尺寸无效: %ux%u step=%u", msg->width, msg->height, msg->step);
      return;
    }
    const auto now = std::chrono::steady_clock::now();
    if (!calibration_mode_ && recognition_rate_ > 0.0 && has_last_publish_) {
      const auto minimum = std::chrono::duration<double>(1.0 / recognition_rate_);
      if (now - last_publish_ < minimum) {
        return;
      }
    }
    last_publish_ = now;
    has_last_publish_ = true;

    const std::size_t half_width = msg->width / 2;
    if (msg->step % msg->width != 0) {
      RCLCPP_ERROR_THROTTLE(
        get_logger(), *get_clock(), 2000,
        "拼接图 step 不能整除 width，无法安全拆分");
      return;
    }
    const std::size_t bytes_per_pixel = msg->step / msg->width;
    const std::size_t half_step = half_width * bytes_per_pixel;
    const std::size_t required = static_cast<std::size_t>(msg->step) * msg->height;
    if (msg->data.size() < required) {
      RCLCPP_ERROR_THROTTLE(get_logger(), *get_clock(), 2000, "拼接图数据长度不足");
      return;
    }

    auto first = make_half_image(*msg, 0, half_width, half_step);
    auto second = make_half_image(*msg, half_step, half_width, half_step);
    sensor_msgs::msg::Image left;
    sensor_msgs::msg::Image right;
    if (left_first_) {
      left = std::move(first);
      right = std::move(second);
    } else {
      left = std::move(second);
      right = std::move(first);
    }
    left.header.frame_id = left_frame_;
    right.header.frame_id = right_frame_;
    auto left_info = left_camera_info_->getCameraInfo();
    auto right_info = right_camera_info_->getCameraInfo();
    left_info.header.stamp = msg->header.stamp;
    right_info.header.stamp = msg->header.stamp;
    left_info.header.frame_id = left_frame_;
    right_info.header.frame_id = right_frame_;
    left_image_pub_->publish(std::move(left));
    right_image_pub_->publish(std::move(right));
    left_info_pub_->publish(std::move(left_info));
    right_info_pub_->publish(std::move(right_info));
  }

  sensor_msgs::msg::Image make_half_image(
    const sensor_msgs::msg::Image & source, std::size_t source_offset,
    std::size_t width, std::size_t half_step) const
  {
    sensor_msgs::msg::Image output;
    output.header = source.header;
    output.height = source.height;
    output.width = static_cast<std::uint32_t>(width);
    output.encoding = source.encoding;
    output.is_bigendian = source.is_bigendian;
    output.step = static_cast<std::uint32_t>(half_step);
    output.data.resize(half_step * source.height);
    for (std::size_t row = 0; row < source.height; ++row) {
      const auto * begin = source.data.data() + row * source.step + source_offset;
      std::memcpy(output.data.data() + row * half_step, begin, half_step);
    }
    return output;
  }

  std::string input_topic_;
  std::string left_topic_;
  std::string right_topic_;
  std::string left_info_topic_;
  std::string right_info_topic_;
  std::string left_frame_;
  std::string right_frame_;
  bool left_first_{true};
  bool calibration_mode_{false};
  double recognition_rate_{10.0};
  bool has_last_publish_{false};
  std::chrono::steady_clock::time_point last_publish_;
  std::unique_ptr<camera_info_manager::CameraInfoManager> left_camera_info_;
  std::unique_ptr<camera_info_manager::CameraInfoManager> right_camera_info_;
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr left_image_pub_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr right_image_pub_;
  rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr left_info_pub_;
  rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr right_info_pub_;
};

}  // namespace robot_stereo_components

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<robot_stereo_components::StereoSplitterNode>());
  rclcpp::shutdown();
  return 0;
}
